telegram/src/main/kotlin/com/github/kotlintelegrambot/dispatcher/Dispatcher.kt

```diff
@@ -7,35 +7,40 @@ import com.github.kotlintelegrambot.entities.Update
 import com.github.kotlintelegrambot.errors.TelegramError
 import com.github.kotlintelegrambot.logging.LogLevel
 import com.github.kotlintelegrambot.types.DispatchableObject
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.yield
 import java.util.concurrent.BlockingQueue
-import java.util.concurrent.Executor
 
 class Dispatcher internal constructor(
     private val updatesQueue: BlockingQueue<DispatchableObject>,
-    private val updatesExecutor: Executor,
     private val logLevel: LogLevel,
+    ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
```

Comment:
I wonder if we should be using `Dispatchers.IO` here. The dispatcher here is used to launch a single coroutine that will loop until stopped, checking for updates on a queue, and running functions provided by consumers of the library for each of those (in those functions consumers can decide whether they need to run in a different dispatcher or not). But we don't know whether those functions will be blocking IO operations so it doesn't feel right to use this `Dispatcher` here. The main downside I could see of this is reducing the potential for parallelism of `Dispatchers.IO`, which is global to the app, because we have one of its threads always busy. In practice, I'm not sure whether this will actually make a difference but it may for some use cases? Any thoughts on this? One alternative could be to just have a dispatcher backed by a `singleThreadExecutor` that we create only for this.

A bit unrelated. This also makes me wonder about next steps for better coroutine adoption, and feels like replacing `BlockingQueue` for a `Channel` could be a nice impovement, so we suspend while waiting for new updates rather than blocking. Also out of scope of this PR, but just bringing it up to get your thoughts about it.

Reply:
`Dispatcher` is a bit different than `Thread`, A `Dispatcher` is more like a `ThreadPool` where coroutines are executed (Except the `Dispatcher.Main` on Android implementation that use the UI Thread`), so it shouldn`t be blocking IO Operations on the rest of the app. Even the coroutine we launch on this Dispatcher will be running indefinitely until the loop is stopped, on every iteration of our infinite loop we are "suspending" our logic and delegating to the dispatcher we can continue (our job has not been canceled) or if there are others coroutines waiting to be executed and need to be run before us. We achieve it by calling `yield()` method.

Apart from that, I had a doubt about using `Dispatchers.Default` or `Dispatchers.IO`. Reading the [docs](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-dispatcher/) First one is used to run "hard jobs" that consume machine recurses and the second one to operations that will be interacting with IO Resources. In our case, I think the IO definition is closer to our implementation (We run IO calls on every iteration). Maybe, something we could do is run `updates` on `Dispatchers.Default` but switch to `Dispatchers.IO` when API calls are executed. For that would be better to improve the "Retrofit integration" to use coroutines internally, but I think it can be achieved in a separate PR if we decide it is needed.

Lastly, regarding the `BlockingQueue`, of course, we could use `Channel` instead, I didn't think about that, but it is a really good point

Reply:
I will try to do separate PR using `Channel` instead of `BlockingQueue`, because If I change it here, lot of code will need to be updated (Tests are using fake blockingQueue that will need to be updated as well..)

Reply:
Yes, right. I see the point of this code being "cooperative" by doing `yield` every time an item is received from the queue, so it's not going to be taking a thread forever. For now, it will only be blocking in `queue.take` but that will become suspending when we migrate to use `Channel` so I think this class is doing good.

But I don't think we run any IO operation in `Dispatcher` (we run them on `Updater`) so that's why I was thinking that maybe it would be nicer to use our own dispatcher backed by a single thread executor so that we don't take any of the resources from `Dispatchers.IO` here. There's this nice extension function to make our work easy https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/as-coroutine-dispatcher.html.

I agree it could make sense to use `Dispatchers.IO` on the `Updater` class, which is where we actually run the `getUpdates` API operation.

Re. Retrofit it might be nice to use the suspend version of the functions but as far as I understand OkHttp doesn't support non-blocking IO operations yet (https://github.com/square/okio/pull/531). One alternative would be to use Ktor instead https://github.com/ktorio/ktor, but that would be a big task, so I guess we can do with execution blocking IO operations on `Dispatchers.IO` for now.

Reply:
Ok, I see. Let me create our own `Dispatcher` and inject it, instead of injecting the `Dispatcher.IO` 

Re Network Layer, I think it is out of the scope of this PR, it can be covered later

Reply:
Yes, completely agree.